x/oracle/keeper/validatorWhiteList.go

```diff
@@ -1,53 +1,78 @@
+	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
-func (k Keeper) SetOracleWhiteList(ctx sdk.Context, validatorList []sdk.ValAddress) {
+// GetAllWhiteList set the validator list for a network.
+func (k Keeper) GetAllWhiteList(ctx sdk.Context) map[types.NetworkID]types.ValidatorWhiteList {
+	result := make(map[types.NetworkID]types.ValidatorWhiteList)
 	store := ctx.KVStore(k.storeKey)
-	key := types.WhiteListValidatorPrefix
-	valList := make([]string, len(validatorList))
-	for i, entry := range validatorList {
-		valList[i] = entry.String()
+	iterator := sdk.KVStorePrefixIterator(store, types.WhiteListValidatorPrefix)
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		networkDescriptor, err := types.GetFromPrefix(iterator.Key())
+		if err != nil {
+			panic(err.Error())
+		}
+
+		result[networkDescriptor.NetworkID] = k.GetOracleWhiteList(ctx, networkDescriptor)
 	}
-	store.Set(key, k.cdc.MustMarshalBinaryBare(&stakingtypes.ValAddresses{Addresses: valList}))
+
+	return result
+}
+
+// SetOracleWhiteList set the validator list for a network.
+func (k Keeper) SetOracleWhiteList(ctx sdk.Context, networkDescriptor types.NetworkDescriptor, validatorList types.ValidatorWhiteList) {
+	store := ctx.KVStore(k.storeKey)
+	key := networkDescriptor.GetPrefix()
+	store.Set(key, k.cdc.MustMarshalBinaryBare(&validatorList))
 }
 
-func (k Keeper) ExistsOracleWhiteList(ctx sdk.Context) bool {
-	key := types.WhiteListValidatorPrefix
+// ExistsOracleWhiteList check if the key exist
+func (k Keeper) ExistsOracleWhiteList(ctx sdk.Context, networkDescriptor types.NetworkDescriptor) bool {
+	key := networkDescriptor.GetPrefix()
 	return k.Exists(ctx, key)
 }
 
-//
-func (k Keeper) GetOracleWhiteList(ctx sdk.Context) []sdk.ValAddress {
+// GetOracleWhiteList return validator list
+func (k Keeper) GetOracleWhiteList(ctx sdk.Context, networkDescriptor types.NetworkDescriptor) types.ValidatorWhiteList {
 	store := ctx.KVStore(k.storeKey)
-	key := types.WhiteListValidatorPrefix
+	// key := types.WhiteListValidatorPrefix
+	key := networkDescriptor.GetPrefix()
 	bz := store.Get(key)
-	valAddresses := &stakingtypes.ValAddresses{}
-	k.cdc.MustUnmarshalBinaryBare(bz, valAddresses)
+	validators := &types.ValidatorWhiteList{}
+	k.cdc.MustUnmarshalBinaryBare(bz, validators)
+	return *validators
+}
+
+// GetAllValidators return validator list
+func (k Keeper) GetAllValidators(ctx sdk.Context, networkDescriptor types.NetworkDescriptor) []sdk.ValAddress {
+	valAddresses := k.GetOracleWhiteList(ctx, networkDescriptor)
 
-	vl := make([]sdk.ValAddress, len(valAddresses.Addresses))
-	for i, entry := range valAddresses.Addresses {
-		addr, err := sdk.ValAddressFromBech32(entry)
+	vl := make([]sdk.ValAddress, len(valAddresses.GetWhiteList()))
+	index := 0
+	for i, power := range valAddresses.GetWhiteList() {
+		addr, err := sdk.ValAddressFromBech32(i)
 		if err != nil {
 			panic(err)
 		}
-		vl[i] = addr
+		if power > 0 {
+			vl[index] = addr
```

Comment:
Here we should be appending to a dynamic array instead of writing directly to an index.